File: 1091-shortest-path-in-binary-matrix/1091-shortest-path-in-binary-matrix.py
class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        if grid[0][0]==1 or grid[-1][-1]==1:
            return -1
        moves = [(1,0),(0,1),(-1,0),(0,-1),(1,1),(-1,-1),(-1,1),(1,-1)]
        n = len(grid)
        visited = [[-1 for _ in range(n)] for i in range(n)]
        visited[0][0],que = 1,[(0,0)]
        while que:
            node = que.pop(0)
            for i in range(8):
                if 0<=node[0]+moves[i][0]<n and 0<=node[1]+moves[i][1]<n and grid[node[0]+moves[i][0]][node[1]+moves[i][1]]!=1 and visited[node[0]+moves[i][0]][node[1]+moves[i][1]]==-1:
                    visited[node[0]+moves[i][0]][node[1]+moves[i][1]] = visited[node[0]][node[1]] + 1
                    que.append((node[0]+moves[i][0],node[1]+moves[i][1]))
        return visited[n-1][n-1]

File: 1091-shortest-path-in-binary-matrix/1091-shortest-path-in-binary-matrix.java
class Solution {
    int[][] directions = {{0,1},{1,0},{-1,0},{0,-1},{1,1},{-1,-1},{1,-1},{-1,1}};
    public int shortestPathBinaryMatrix(int[][] grid) {
        if(grid[0][0]==1){
            return -1;
        }
        Queue<int[]> que = new LinkedList<int[]>();
        int[][] direction = new int[][]{{0,1},{1,0},{0,-1},{-1,0},{1,1},{-1,-1},{1,-1},{-1,1}};
        que.add(new int[]{0,0,1});
        int n = grid.length;
        int m = grid[0].length;
        int[][] visited = new int[n][m];
        visited[0][0] = 1;
        while(!que.isEmpty()){
            int k = que.size();
            for(int i=0;i<k;i++){
                int[] node = que.poll();
                if(node[0]==n-1 && node[1]==m-1){
                    return node[2];
                }
                for(int j=0;j<8;j++){
                    int newX = node[0] + direction[j][0];
                    int newY = node[1] + direction[j][1];
                    int distance = node[2];
                    if(newX>=0 && newX<n && newY>=0 && newY<m && visited[newX][newY]!=1 && grid[newX][newY]==0){
                        visited[newX][newY] = 1;
                        que.add(new int[]{newX,newY,distance+1});
                    }
                }
            }
        }
        return -1;
    }
}